### PR TITLE
SSL improvements

### PR DESCRIPTION
**Pull request by vietj**
A set of improvements for SSL handling

- SSL configuration is performed
  - on a worker thread instead of event-loop thread
  - lazily in `NetClient` as validation is asynchronous
- `SslContext` is now delegated to an SPI to enable alternative way to create an `SslContext` (e.g reuse an existing `SSLContext`) provided on the `SSLEngineOptions` classes
- remove the _suppliedSSLContext_ temporary machinery in favor of the `SSLEngineOptions` configuration mechanism
- _ssl_ handler (`SslHandler` or `SniHandler`) is now created by the `SSLHelper` instead of the pipeline setup
- internal cleanups

The SPI will be documented in the [Vert.x Advanced Guide](https://github.com/vert-x3/advanced-vertx-guide)

Thanks @MikeEdgar for the SSL blocking contribution

### CODE REVIEW

**Comment by vietj** [comment]
@kabir have a look, it changes the temporary supplied SSLContext machinery for a cleaner API, `NetTest#testNetClientInternalTLS` shows how the API can be used for `NetClient`. I will provide a PR for vertx-proton for this matter.

**Comment by kabir** [comment]
@vietj Looks good. My only concern is what the timeframe is to get this in? Since I have some work further down the chain (in SR Reactive Messaging) which needs merging/adjusting.

CC @cescoffier @ozangunalp

**Comment by vietj** [comment]
I think the next release. I'm working actively on it.

**Comment by kabir** [comment]
@vietj ok, great - the needed work in SmallRye Reactive Messaging will need updating. But since that isn't merged yet, assuming this comes soon-ish, I can update that PR.

**Comment by MikeEdgar** [comment]
@vietj - this looks good to me as it relates to issue #4439. I no longer see the blocked threads during application startup with this branch.

**Comment by vietj** [comment]
@kabir this is I think the most up to date PR for the feature

Now the vertx-proton and vertx-amqp-client changes are not necessary anymore as the SSLContext can be configured from the `SSLEngineOptions`, e.g

```java
    AmqpClientOptions options = new AmqpClientOptions()
      .setSsl(true)
      .setHost("localhost")
      .setPort(server.actualPort())
      .setSslEngineOptions(new JdkSSLEngineOptions() {
        @Override
        public SslContextFactory sslContextFactory() {
          return new SslContextFactory() {
            @Override
            public SslContext create() throws SSLException {
              return new JdkSslContext(
                sslContext,
                true,
                null,
                IdentityCipherSuiteFilter.INSTANCE,
                ApplicationProtocolConfig.DISABLED,
                io.netty.handler.ssl.ClientAuth.NONE,
                null,
                false);
            }
          };
        }
      });

    AmqpClient client = AmqpClient.create(vertx, options);
```

After this is merged, I'll revert the changes in vertx-proton and vertx-amqp-client (but I'll keep the SSL tests).

**Comment by kabir** [comment]
Thanks @vietj - the changes look good to me

File: src/main/java/io/vertx/core/net/impl/NetClientImpl.java [hunk 1]
```diff
@@ -226,6 +212,26 @@ private void connect(SocketAddress remoteAddress, String serverName, Promise<Net
   }
 
   public void connectInternal(ProxyOptions proxyOptions,
```
Comment (InfoSec812) [inline src/main/java/io/vertx/core/net/impl/NetClientImpl.java:214]: Might be helpful to add JavaDoc to this method and the next?

File: src/main/java/io/vertx/core/net/impl/TCPServerBase.java [hunk 2]
```diff
@@ -96,21 +91,10 @@ public synchronized SSLHelper sslHelper() {
...
-      } else {
-        promise.fail(res.cause());
-      }
-    });
-    return promise.future();
+    return listen(address, listenContext).map(this);
   }
 
-  private synchronized io.netty.util.concurrent.Future<Channel> listen(SocketAddress localAddress, ContextInternal context) {
+  private synchronized Future<Channel> listen(SocketAddress localAddress, ContextInternal context) {
```
Comment (InfoSec812) [inline src/main/java/io/vertx/core/net/impl/TCPServerBase.java:97]: I am having issues tracing through this method and understanding the reasoning. Any chance of adding comments or decomposing into smaller methods?

**Review by InfoSec812 — COMMENTED** [review]
I added some comments from my perspective as a user of the toolkit who is trying to learn more about the internals.

File: src/main/java/io/vertx/core/net/impl/TCPServerBase.java (hunk 2 above)
Comment (vietj) [reply src/main/java/io/vertx/core/net/impl/TCPServerBase.java:97]: I'll add comments

File: src/main/java/io/vertx/core/net/impl/NetClientImpl.java (hunk 1 above)
Comment (vietj) [reply src/main/java/io/vertx/core/net/impl/NetClientImpl.java:214]: will do

**Review by InfoSec812 — APPROVED** [review]
Thanks for the additional comments and documentation @vietj - Looks good!

**Review by tsegismont — APPROVED** [review]
@vietj great job

**Pull request merged**